protocol/messenger_handler.go

```diff
@@ -406,20 +412,114 @@ func (m *Messenger) handleCommandMessage(state *ReceivedMessageState, message *c
 	return nil
 }
 
-func (m *Messenger) HandleBackup(state *ReceivedMessageState, message protobuf.Backup) error {
+func (m *Messenger) HandleBackup(state *ReceivedMessageState, message protobuf.Backup) []error {
+	var errors []error
+	err := m.handleSyncProfile(state, message.Profile, message.Clock)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
 	for _, contact := range message.Contacts {
-		err := m.HandleSyncInstallationContact(state, *contact)
+		err = m.HandleSyncInstallationContact(state, *contact)
 		if err != nil {
-			return err
+			errors = append(errors, err)
 		}
 	}
 
 	for _, community := range message.Communities {
-		err := m.handleSyncCommunity(state, *community)
+		err = m.handleSyncCommunity(state, *community)
 		if err != nil {
-			return err
+			errors = append(errors, err)
 		}
 	}
+	err = m.handleSyncSettings(state, message.Settings)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	if m.config.messengerSignalsHandler != nil {
+		response := wakusync.SyncFromWakuResponse{}
+
+		response.AddWakuSyncDetails(SyncWakuSectionKeyProfile, message.ProfileDetails)
+		response.AddWakuSyncDetails(SyncWakuSectionKeyContacts, message.ContactsDetails)
+		response.AddWakuSyncDetails(SyncWakuSectionKeyCommunities, message.CommunitiesDetails)
+		response.AddWakuSyncDetails(SyncWakuSectionKeySettings, message.SettingsDetails)
+
+		m.config.messengerSignalsHandler.SendSyncFromWakuProgress(&response)
+	}
+
+	return errors
+}
+
+func (m *Messenger) handleSyncProfile(state *ReceivedMessageState, message *protobuf.SyncProfile, backupTime uint64) error {
+	if message == nil {
+		return nil
+	}
+
+	pubKey := common.PubkeyToHex(&m.identity.PublicKey)
+	contact, ok := state.AllContacts.Load(pubKey)
+	if !ok {
+		return errors.New("cannot find logged in profile among all known contacts")
+	}
+
+	response := wakusync.SyncFromWakuResponse{
+		Profile: &wakusync.SyncProfile{},
+	}
+
+	if contact.LastUpdated < backupTime {
```

Comment:
BackupTime is not the clock, since that's only the clock of the time you sent a backup, but it might not reflect the latest user's update. You need to have a clock for each piece that syncs independently.

Reply:
`Clock` value of `protobuf.Backup` message is actually `clock` value got from this call `clock, chat := m.getLastClockWithRelatedChat()` is that wrong?

Reply:
@cammellos great example, thanks, explains everything, will update the code accordingly.